ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/PullProjectNode.java

```diff
@@ -0,0 +1,260 @@
+import io.confluent.ksql.schema.ksql.LogicalSchema;
+import io.confluent.ksql.schema.ksql.LogicalSchema.Builder;
+import io.confluent.ksql.schema.ksql.SystemColumns;
+import io.confluent.ksql.schema.ksql.types.SqlType;
+import io.confluent.ksql.schema.ksql.types.SqlTypes;
+import io.confluent.ksql.util.KsqlConfig;
+import io.confluent.ksql.util.KsqlException;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * The projection of a Pull query.
+ *
+ * <p>There are 3 schemas the node is handling, the input schema, the intermediate schema used
+ * for codegen and the output schema.
+ * <ul>
+ * <li>The input is the schema of the child node
+ *
+ * <li>CodeGen is used only if the projection is not SELECT *. For CodeGen, an intermediate schema
+ * is created as follows:
+ * Check if projection contains system or key columns. If not, the intermediate schema
+ * is the input schema. If there are any of these columns, the input schema is extended by copying
+ * the key and system columns (rowtime, windwostart and windowend) into the value of the schema.
+ *
+ * <li>For the output schema, if the projection is SELECT *, add windowstart and windowend to key
+ * columns and keep value columns the same as input. If projection is not SELECT *,
+ * then process each select and if it is a key or windowstart and windowend add them to the key
+ * part else add them to the value part.
+ * </ul>
+ */
+public class PullProjectNode extends ProjectNode {
+
+  private final Projection projection;
+  private final ImmutableList<SelectExpression> selectExpressions;
+  private final LogicalSchema outputSchema;
+  private final LogicalSchema intermediateSchema;
+  private final List<ExpressionMetadata> compiledSelectExpressions;
+  private final RewrittenAnalysis analysis;
+  private final boolean isSelectStar;
+  private final boolean addAdditionalColumnsToIntermediateSchema;
+
+  public PullProjectNode(
+      final PlanNodeId id,
+      final PlanNode source,
+      final List<SelectItem> selectItems,
+      final MetaStore metaStore,
+      final KsqlConfig ksqlConfig,
+      final RewrittenAnalysis analysis
+  ) {
+    super(id, source);
+    this.projection = Projection.of(selectItems);
+    this.analysis = Objects.requireNonNull(analysis, "analysis");
+    this.selectExpressions = ImmutableList.copyOf(SelectionUtil
+        .buildSelectExpressions(getSource(), projection.selectItems(), Optional.empty()));
+    this.isSelectStar = isSelectStar();
+    this.addAdditionalColumnsToIntermediateSchema = shouldAddAdditionalColumnsInSchema();
+    this.outputSchema = buildOutputSchema(metaStore);
+    this.intermediateSchema = buildIntermediateSchema();
+    this.compiledSelectExpressions = selectExpressions
+        .stream()
+        .map(selectExpression -> CodeGenRunner.compileExpression(
+            selectExpression.getExpression(),
+            "Select",
+            intermediateSchema,
```

Comment:
Yeah, for the next PR that does the WHERE clause I would have to share this schema. I will build it once in the WHERE and then the PROJECT since it sits above, will be able to get if from it. If a WHERE doesn't exist, then the PROJECT will build it